build-system/tasks/runtime-test/runtime-test-base.js

```diff
@@ -0,0 +1,249 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+const argv = require('minimist')(process.argv.slice(2));
+const babelify = require('babelify');
+const karmaConfig = require('../karma.conf');
+const log = require('fancy-log');
+const {
+  createKarmaServer,
+  getAdTypes,
+  runTestInBatches,
+  startTestServer,
+} = require('./helpers');
+const {createCtrlcHandler, exitCtrlcHandler} = require('../../ctrlcHandler');
+const {green, yellow, cyan, red} = require('ansi-colors');
+const {isTravisBuild} = require('../../travis');
+const {reportTestStarted} = require('.././report-test-status');
+
+class RuntimeTestConfig {
+  constructor(testType) {
+    this.testType = testType;
+
+    Object.assign(this, karmaConfig);
+    this.browsers = this.getBrowsersObject().browsers;
+    this.files = this.getFiles();
+    this.reporters = this.getReporters();
+    this.singleRun = !argv.watch && !argv.w;
+    this.client.mocha.grep = !!argv.grep;
+    this.client.verboseLogging = !!argv.verbose || !!argv.v;
+    this.client.captureConsole = !!argv.verbose || !!argv.v || !!argv.files;
+    this.browserify.configure = function(bundle) {
+      bundle.on('prebundle', function() {
+        log(
+          green('Transforming tests with'),
+          cyan('browserify') + green('...')
+        );
+      });
+      bundle.on('transform', function(tr) {
+        if (tr instanceof babelify) {
+          tr.once('babelify', function() {
+            process.stdout.write('.');
+          });
+        }
+      });
+    };
+
+    // c.client is available in test browser via window.parent.karma.config
+    this.client.amp = {
+      useCompiledJs: !!argv.compiled,
+      saucelabs: !!argv.saucelabs,
+      singlePass: !!argv.single_pass,
+      adTypes: getAdTypes(),
+      mochaTimeout: this.client.mocha.timeout,
+      propertiesObfuscated: !!argv.single_pass,
+      testServerPort: this.client.testServerPort,
+      testOnIe:
+        this.browsers.includes('IE') || this.browsers.includes('SL_IE_11'),
+    };
+
+    if (argv.coverage && this.testType != 'a4a') {
+      this.plugins.push('karma-coverage-istanbul-reporter');
+      this.coverageIstanbulReporter = {
+        dir: 'test/coverage',
+        reports: isTravisBuild()
+          ? ['lcovonly']
+          : ['html', 'text', 'text-summary'],
+        'report-config': {lcovonly: {file: `lcov-${testType}.info`}},
+      };
+
+      const plugin = [
+        'istanbul',
+        {
+          exclude: [
+            'ads/**/*.js',
+            'third_party/**/*.js',
+            'test/**/*.js',
+            'extensions/**/test/**/*.js',
+            'testing/**/*.js',
+          ],
+        },
+      ];
+
+      this.browserify.transform = [['babelify', {plugins: [plugin]}]];
+    }
+  }
+
+  getBrowsersObject() {
```

Comment:
does it feel a bit awkward to you that this is a `config` instance but is not necessarily a value type? (it has methods on it etc). I don't feel too strongly but i prefer my config instances to be purely a value type in case i need to serialize it

Reply:
Good point, I agree.

Reply:
feel free to punt though. don't necessarily need to refactor this now

Reply:
sg